.evergreen.yml

```diff
@@ -218,115 +226,173 @@ tasks:
+        variant: darwin
+      - name: test_m40x
+        variant: darwin
+      - name: test_m42x
+        variant: darwin
+      - name: test_m44x
+        variant: darwin
     commands:
       - func: checkout
       - func: install
-      - func: release_debian
-  - name: release_rhel
+      - func: compile_and_upload_artifact
+  - name: compile_and_upload_artifact_linux
     depends_on:
       - name: check
-      - name: test
+      - name: test_m40x
+        variant: linux
+      - name: test_m42x
+        variant: linux
+      - name: test_m44x
         variant: linux
     commands:
       - func: checkout
       - func: install
-      - func: release_rhel
-  - name: release_win_ps
+      - func: compile_and_upload_artifact
+  - name: compile_and_upload_artifact_ps
     depends_on:
       - name: check_ps
+      - name: test_m40x_ps
+      - name: test_m42x_ps
+      - name: test_m44x_ps
     commands:
       - func: checkout
       - func: install_ps
-      - func: release_win_ps
+      - func: compile_and_upload_artifact_ps
+  - name: test_linux_artifact
+    depends_on:
+      - name: compile_and_upload_artifact_linux
+        variant: linux
+    commands:
+      - func: checkout
+      - func: install
+      - func: test_linux_artifact
   - name: release_publish
     depends_on:
-      - name: release_win_ps
+      - name: compile_and_upload_artifact_ps
         variant: win32
-      - name: release_linux
+      - name: compile_and_upload_artifact_linux
+        variant: linux
+      - name: test_linux_artifact
+        variant: rhel_real
+      - name: test_linux_artifact
+        variant: debian_real
+      - name: compile_and_upload_artifact_macos
+        variant: darwin_codesign
+      - name: test_vscode
         variant: linux
-      - name: release_rhel
-        variant: rhel
-      - name: release_debian
-        variant: debian
-      - name: release_macos
-        variant: darwin
     commands:
       - func: checkout
       - func: install
       - func: release_publish
 
 # Need to run builds for every possible build variant.
 buildvariants:
-  - name: darwin
-    display_name: "MacOS Mohave"
+  - name: darwin_codesign
+    display_name: "MacOS Mojave (codesign)"
     run_on: macos-1014-codesign
+    tasks:
+      - name: compile_and_upload_artifact_macos
+  - name: darwin
+    display_name: "MacOS Mojave"
+    run_on: macos-1014
     tasks:
       - name: check
-      - name: test
-      - name: release_macos
+      - name: test_m40x
+      - name: test_m42x
+      - name: test_m44x
   - name: linux
     display_name: "Ubuntu 18.04"
     run_on: ubuntu1804-test
     tasks:
       - name: check
-      - name: test
+      - name: test_m40x
+      - name: test_m42x
+      - name: test_m44x
       - name: test_vscode
-      - name: release_linux
+      - name: compile_and_upload_artifact_linux
   - name: rhel
-    display_name: "RHEL 8.0 (in Ubuntu docker)"
+    display_name: "Ubuntu 18.04 (rpm target)"
```

Comment:
Stupid question: what are `rhel` and `debian`? They are both running on `ubuntu1804-test` and thus the same as `linux`? 🤔 Can we throw them out?

Reply:
Ah, right :slightly_smiling_face: So … our `@mongosh/build` package makes a distinction on how to behave based on the build variant name, and, in particular, it packages as a tarball/deb/rpm based on it. The variable gets picked up in `config/build.conf.js` from an environment variable, and passed around throughout the build package.

That’s why the names here, `rhel` and `debian`, are not freely changeable:

https://github.com/mongodb-js/mongosh/blob/dde7b239f0cef50d217a4bf1766a89c8890a2e9d/packages/build/src/build-variant.ts#L7-L13

This is obviously not an ideal situation, no. It means that we’re currently building the Linux binary three separate times, only differing in the final packaging step, which wastes lots of CPU time in CI :slightly_smiling_face: But I also wanted to get the specific task in question done, and this doesn’t make anything worse in this respect as far as I can tell.

Ultimately, it would be nice imo if:

- We didn’t make choices on how to behave purely based on the build variant name, and
- We only built the Linux binary exactly once.

Reply:
Thank you very much! Didn't expect these names to be relevant outside the config, but that's good to know 👍